test/gtest/ucp/test_ucp_sockaddr.cc

```diff
@@ -721,6 +722,22 @@ class test_ucp_sockaddr_destroy_ep_on_err : public test_ucp_sockaddr {
         e->disconnect_nb(0, 0, UCP_EP_CLOSE_MODE_FORCE);
     }
 
+    static ucs_log_func_rc_t
+    wrap_wireup_ep_not_connected_logger(const char *file, unsigned line, const char *function,
+                                        ucs_log_level_t level,
+                                        const ucs_log_component_config_t *comp_conf,
+                                        const char *message, va_list ap)
+    {
+        if (level == UCS_LOG_LEVEL_WARN) {
+            std::string err_str = format_message(message, ap);
+            if (strstr(err_str.c_str(), "Endpoint is not connected")) {
```

Comment:
why would we have such error?

Reply:
you mean why it's wrapped or why we print a warning on failure to disconnect a cm ep?

Reply:
what is the flow which added that leads to such print
maybe we need to not print it at all

Reply:
please see the flow in the 3rd comment in the commit message:

3. wrap an error from wireup_cm saying that the cm ep failed to
disconnect due to - 'Endpoint is not connected'.
with tcp_sockcm this status is returned from the disconnect function in
case the ep is in the Failed state. The Failed state could be set
following a remote close, like in bullet 2 above.
async thread could get the error event and the main thread could close
the ep right after.

Reply:
does it mean that in this case a regular UCP user could see this error during disconnect? so even though disconnect is successful, there will still be error print?

Reply:
yes, but it's a warning, not an error...

Reply:
warnings are also visible by default, it should not be printed in this case, since it's not really an error. it's just a simultaneous disconnect.

Reply:
you mean to remove this warning print (set it to debug)? 
https://github.com/openucx/ucx/blob/master/src/ucp/wireup/wireup_cm.c#L979

Reply:
can we check the ucp_ep state and avoid calling disconnect in such flow?

Reply:
not sure since the async thread could change the sockaddr ep to Failed before ucp had a chance to see it.

Reply:
check state under lock?

Reply:
if connection was established we have to call disconnect from both sides in any case since local side does not know how remote side was closed, if I understand correctly the issue it should be handled in UCT/TCP/CM

Reply:
tcp sockcm returns NOT_CONNECTED which is the actual case. maybe need to check that this is the status in this line or print, and if so, not to print the warning?

Reply:
why NOT_CONNECTED? why UCT ep is in failed state? due to remote disconnect notification?

Reply:
NOT_CONNECTED is returned from tc_sockcm if ep is in Failed state.
it is in Failed state since the remote peer closed with Force, so the EPIPE event was generated.

Reply:
what's the difference with force or not? locally (in UCP) we should get the same status in disconnect callback, right?

Reply:
this is the error that tcp_sockcm gets when the remote peer 'disappears'.
but this isn't the disconnect callback. this is a print from a call to disconnect a sockaddr cm ep (not as a result of remote disconnect, but from an ep_close_nb flow).

Reply:
but you wrote:
> the async thread could change the sockaddr ep to Failed before ucp had a chance to see it.

so, is there any other events or not?
if this is a first error reported to UCP then it should be CONN_RESET and UCP has to handle this the same way as disconnect callback.

Reply:
what if there was a failure in tcp_sockcm but before the disconnect_cb was called, the main thread already called ep_disconnect?

Reply:
> what if there was a failure in tcp_sockcm but before the disconnect_cb was called, the main thread already called ep_disconnect?

the progress state would not be able to change tcp ep state if main thread holds a still the lock while calling uct_ep_disconnect

Reply:
but ucp registers a function on the main thread to perform the actual disconnect_cb work.
so tcp_sockcm called disconnect_cb, changed the state for Failed, main thread called ep_close before the slow_path_progress function was called.

Reply:
can ucp change the state flags from progress thread and do the rest of the work on main thread?